src/allocation/mod.rs

```diff
@@ -174,12 +180,12 @@ impl Allocation {
     }
 
     // Close closes the allocation
-    pub async fn close(&mut self) -> Result<()> {
-        if self.closed {
+    pub async fn close(&self) -> Result<()> {
+        if self.closed.load(Ordering::Acquire) {
             return Err(Error::ErrClosed);
         }
 
-        self.closed = true;
+        self.closed.store(true, Ordering::Release);
```

Comment:
@logist322 with atomics you do have atomicity regarding the single operation, but you don't have atomicity with two separate operations. So, loading and then storing separately is not atomic and will race. That's why there are compare-and-swap operations on atomics.